src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -107,8 +260,177 @@ def visit_FieldSymbol(
+            case foast.BinaryOperator.BIT_AND:
+                return "&"
+            case foast.BinaryOperator.BIT_OR:
+                return "|"
+
+    def _deduce_arithmetic_binop_type(
+        self,
+        op: foast.BinaryOperator,
+        *,
+        parent: foast.BinOp,
+        left_type: foast.SymbolType,
+        right_type: foast.SymbolType,
+        **kwargs,
+    ) -> foast.SymbolType:
+        left, right = TypeInfo(left_type), TypeInfo(right_type)
+        if (
+            left.is_arithmetic_compatible
+            and right.is_arithmetic_compatible
+            and dims_broadcast_compatible(left, right)
+        ):
+            return dims_broadcasted(left, right).type
+        else:
+            raise FieldOperatorTypeDeductionError.from_foast_node(
+                parent,
+                msg=f"Incompatible type(s) for operator '{self._binop_to_str(op)}': {left.type}, {right.type}!",
+            )
+
+    def _deduce_logical_binop_type(
+        self,
+        op: foast.BinaryOperator,
+        *,
+        parent: foast.BinOp,
+        left_type: foast.SymbolType,
+        right_type: foast.SymbolType,
+        **kwargs,
+    ) -> foast.SymbolType:
+        left, right = TypeInfo(left_type), TypeInfo(right_type)
+        if (
+            left.is_logics_compatible
+            and right.is_logics_compatible
+            and dims_broadcast_compatible(left, right)
+        ):
+            return dims_broadcasted(left, right).type
+        else:
+            raise FieldOperatorTypeDeductionError.from_foast_node(
+                parent,
+                msg=f"Incompatible type(s) for operator '{self._binop_to_str(op)}': {left.type}, {right.type}!",
+            )
+
+    def visit_UnaryOp(self, node: foast.UnaryOp, **kwargs) -> foast.UnaryOp:
+        new_operand = self.visit(node.operand, **kwargs)
+        if not self._unaryop_type_compatible(op=node.op, operand_type=new_operand.type):
+            raise FieldOperatorTypeDeductionError.from_foast_node(
+                node,
+                msg=f"Incompatible type for unary operator '{self._unaryop_to_str(node.op)}': {new_operand.type}!",
+            )
+        return foast.UnaryOp(
+            op=node.op, operand=new_operand, location=node.location, type=new_operand.type
+        )
+
+    def _unaryop_type_compatible(
+        self, op: foast.UnaryOperator, operand_type: foast.FieldType
+    ) -> bool:
+        operand_ti = TypeInfo(operand_type)
+        if op in [foast.UnaryOperator.UADD, foast.UnaryOperator.USUB]:
+            return operand_ti.is_arithmetic_compatible
+        elif op is foast.UnaryOperator.NOT:
+            return operand_ti.is_logics_compatible
+
+    def _unaryop_to_str(self, op: foast.UnaryOperator):
+        match op:
+            case foast.UnaryOperator.UADD:
+                return "+"
+            case foast.UnaryOperator.USUB:
+                return "-"
+            case foast.UnaryOperator.NOT:
+                return "not"
```

Comment:
Missing return type hint. I would additionally move it to `foast.UnaryOperator.__str__`